src/token/ERC20Gauges.sol

```diff
@@ -0,0 +1,372 @@
+            emit IncrementGaugeWeight(msg.sender, gauge, weight);
+        }
+
+        // Ensure under weight
+        newUserWeight = getUserWeight[msg.sender] + weightsSum;
+        if (newUserWeight > balanceOf[msg.sender]) revert OverWeightError();
+
+        // update global state
+        getUserWeight[msg.sender] = newUserWeight; 
+        totalWeight += weightsSum;
+    }
+
+    /** 
+     @notice decrement a gauge with some weight for the caller
+     @param gauge the gauge to decrement
+     @param weight the amount of weight to decrement on gauge
+     @return newUserWeight the new user weight
+    */
+    function decrementGauge(address gauge, uint256 weight) external returns (uint256 newUserWeight) {
+        // All operations will revert on underflow, protecting against bad inputs
+        newUserWeight = getUserWeight[msg.sender] - weight;
+
+        getUserWeight[msg.sender] = newUserWeight;
+        getUserGaugeWeight[msg.sender][gauge] -= weight;
+        getGaugeWeight[gauge] -= weight;
+        totalWeight -= weight;
+
+        emit DecrementGaugeWeight(msg.sender, gauge, weight);
+    }
+
+    /** 
+     @notice decrement a list of gauges with some weights for the caller
+     @param gaugeList the gauges to decrement
+     @param weights the list of weights to decrement on the gauges
+     @return newUserWeight the new user weight
+    */
+    function decrementGauges(address[] calldata gaugeList, uint256[] calldata weights) external returns (uint256 newUserWeight) {
+        uint256 size = gaugeList.length;
+        if (weights.length != size) revert SizeMismatchError();
+
+        // store total in summary for batch update on user/global state
+        uint256 weightsSum;
+
+        // Update gauge specific state
+        // All operations will revert on underflow, protecting against bad inputs
+        for (uint256 i = 0; i < size; i++) {
+            address gauge = gaugeList[i];
+            uint256 weight = weights[i];
+            weightsSum += weight;
+
+            getUserGaugeWeight[msg.sender][gauge] -= weight;
+            getGaugeWeight[gauge] -= weight;
+            emit DecrementGaugeWeight(msg.sender, gauge, weight);
+        }
+
+        // update global state
+        newUserWeight = getUserWeight[msg.sender] - weightsSum;
+
+        getUserWeight[msg.sender] = newUserWeight;
+        totalWeight -= weightsSum;
+    }
+
+    /// @notice free deprecated gauges for a user. This method can be called by anyone.
+    function freeDeprecatedGauges(address user, address[] memory gaugeList) public {
+        uint256 size = gaugeList.length;
+
+        uint256 totalFree;
+        for (uint256 i = 0; i < size; i++) {
+            address gauge = gaugeList[i];
+            if (!_deprecatedGauges.contains(gauge)) revert InvalidGaugeError();
+
+            uint256 weight = getUserGaugeWeight[user][gauge];
+            if (weight != 0) {
+                getUserGaugeWeight[user][gauge] = 0;
```

Comment:
delete instead of assign to 0.

Reply:
syntactically those are the same. Delete is most helpful for arrays/structs not basic types